File: stacks/go-gin/examples/blog-api/internal/service/post_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/example/blog-api/internal/model"
)

// ErrPostNotFound is returned when a post does not exist.
var ErrPostNotFound = errors.New("post not found")

// PostRepository is the persistence contract this service depends on.
type PostRepository interface {
	Create(ctx context.Context, post *model.Post) error
	GetByID(ctx context.Context, id string) (*model.Post, error)
	List(ctx context.Context, limit, offset int) ([]model.Post, int64, error)
	Update(ctx context.Context, post *model.Post) error
	Delete(ctx context.Context, id string) error
}

// PostService contains business logic for blog post operations.
type PostService struct {
	repo PostRepository
}

// NewPostService constructs a PostService with the given repository.
func NewPostService(repo PostRepository) *PostService {
	return &PostService{repo: repo}
}

// CreatePost validates and persists a new post.
func (s *PostService) CreatePost(ctx context.Context, title, content, authorID string) (*model.Post, error) {
	post := &model.Post{Title: title, Content: content, AuthorID: authorID}
	if err := s.repo.Create(ctx, post); err != nil {
		return nil, fmt.Errorf("post service: create: %w", err)
	}
	return post, nil
}

// GetPostByID retrieves a post by ID.
//
// It returns ErrPostNotFound if no post matches.
func (s *PostService) GetPostByID(ctx context.Context, id string) (*model.Post, error) {
	post, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, ErrPostNotFound
	}
	return post, nil
}

// ListPosts returns a page of posts and the total count.
func (s *PostService) ListPosts(ctx context.Context, limit, offset int) ([]model.Post, int64, error) {
	posts, total, err := s.repo.List(ctx, limit, offset)
	if err != nil {
		return nil, 0, fmt.Errorf("post service: list: %w", err)
	}
	return posts, total, nil
}

// UpdatePost applies partial updates to an existing post.
//
// It returns ErrPostNotFound if no post matches the given ID.
func (s *PostService) UpdatePost(ctx context.Context, id string, title, content *string) (*model.Post, error) {
	post, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, ErrPostNotFound
	}

	if title != nil {
		post.Title = *title
	}
	if content != nil {
		post.Content = *content
	}

	if err := s.repo.Update(ctx, post); err != nil {
		return nil, fmt.Errorf("post service: update: %w", err)
	}
	return post, nil
}

// DeletePost removes a post by ID.
//
// It returns ErrPostNotFound if no post matches the given ID.
func (s *PostService) DeletePost(ctx context.Context, id string) error {
	if _, err := s.repo.GetByID(ctx, id); err != nil {
		return ErrPostNotFound
	}
	if err := s.repo.Delete(ctx, id); err != nil {
		return fmt.Errorf("post service: delete: %w", err)
	}
	return nil
}

File: stacks/node-express/examples/blog-api/src/services/post.service.ts
import { PostRepository } from '../repositories/post.repository';
import { Post } from '../schemas/post.schema';
import { NotFoundError, ValidationError } from '../errors';
import { logger } from '../config/logger';

export class PostService {
  constructor(private readonly postRepo: PostRepository) {}

  /**
   * Create a new blog post.
   * @throws {ValidationError} If the title is too short for a meaningful post.
   */
  async createPost(title: string, content: string, authorId: string): Promise<Post> {
    if (title.trim().length < 3) {
      throw new ValidationError('Title must be at least 3 characters');
    }
    try {
      const post = await this.postRepo.create({ title: title.trim(), content, authorId });
      logger.info({ postId: post.postId, authorId }, 'Post created');
      return post;
    } catch (error) {
      logger.error({ err: error }, 'Error creating post');
      throw error;
    }
  }

  /**
   * Get a post by ID.
   * @throws {NotFoundError} If the post does not exist.
   */
  async getPostById(postId: string): Promise<Post> {
    const post = await this.postRepo.getById(postId);
    if (!post) {
      throw new NotFoundError('Post not found');
    }
    return post;
  }

  /** List posts with pagination. */
  async listPosts(limit: number, offset: number): Promise<{ posts: Post[]; totalCount: number; hasMore: boolean }> {
    const { posts, totalCount } = await this.postRepo.getAll(limit, offset);
    return { posts, totalCount, hasMore: offset + posts.length < totalCount };
  }

  /**
   * Update a post. Only the author may update their own post.
   * @throws {NotFoundError} If the post does not exist.
   */
  async updatePost(
    postId: string,
    authorId: string,
    data: { title?: string; content?: string }
  ): Promise<Post> {
    const existing = await this.getPostById(postId);
    if (existing.authorId !== authorId) {
      throw new NotFoundError('Post not found');
    }
    return this.postRepo.update(postId, data);
  }

  /**
   * Delete a post. Only the author may delete their own post.
   * @throws {NotFoundError} If the post does not exist or belongs to another author.
   */
  async deletePost(postId: string, authorId: string): Promise<void> {
    const existing = await this.getPostById(postId);
    if (existing.authorId !== authorId) {
      throw new NotFoundError('Post not found');
    }
    await this.postRepo.delete(postId);
    logger.info({ postId }, 'Post deleted');
  }
}
